symbols/arm7.yml

```diff
@@ -1,88 +1,153 @@
 arm7:
   versions:
-    - EU
-    - NA
-    - JP
+    - EU-ENTRY
+    - NA-ENTRY
+    - JP-ENTRY
+    - EU-WRAM
+    - NA-WRAM
+    - JP-WRAM
+    - EU-SLOW
+    - NA-SLOW
+    - JP-SLOW
   address:
-    EU: 0x2380000
-    NA: 0x2380000
-    JP: 0x2380000
+    EU-ENTRY: 0x2380000
+    NA-ENTRY: 0x2380000
+    JP-ENTRY: 0x2380000
+    EU-WRAM: 0x37F8000
+    NA-WRAM: 0x37F8000
+    JP-WRAM: 0x37F8000
+    EU-SLOW: 0x27E0000
```

Comment:
> I haven't had a chance to take a closer look at the 0x27E0000 area since most of the code I've had to work with for arm7 patching has been within the 0x37F8000 area, but given that 0xF608 is just under 64kB which is the amount of WRAM reserved specifically for arm7 (see https://problemkaputt.de/gbatek.htm#dsmemorymaps), and the 0x27E0000 area has a much larger size than the WRAM area, I think this is a good hypothesis for now.

Hmm, wait a second, I realized that this offset looked familiar. A long time ago I had a theory that this was the DTCM region, which I wrote down in my [EoS RAM notes](https://docs.google.com/document/d/1_Q_7BGmNx5wJtJ9iJEwlK1WITjiCcEQxE9C82RECJbg/edit#heading=h.jtqakz6si98g). IIRC, this was based on looking at raw memory in DeSmuME. What do you think of this? If you compare the bytes at 0x27E0000 to those at 0x238F7F0 (0xF7F0 relative to arm7.bin), do they really match for 0x17878 bytes? Or does the correspondence stop after 0x4000 bytes as I mentioned in my old notes?

Reply:
Interesting... let me check. Btw do you know how one can do a full memory dump in no$gba? It always says "Huh?" for me whenever I try to dump the whole thing.

Reply:
I can confirm from the decomp that 0x27E0000 is the DTCM region.

Reply:
Interestingly, no$gba shows that 0x27E0000 contains the arm7 code I expect when I switch to nds7, but when I switch back to nds9 it shows me a different set of values, so I think the two cpus might have different memory mapped there. Although it's weird how it's the same exact address, does arm7 have something akin to DTCM? If this region is just slow memory I'm not sure why the devs picked the exact same place to load code.

Reply:
> Btw do you know how one can do a full memory dump in no$gba? It always says "Huh?" for me whenever I try to dump the whole thing.

Not sure, I don't really use No$ all that much. @End45 would probably know?

> I can confirm from the decomp that 0x27E0000 is the DTCM region.

Thanks. Thinking about this more, it's kind of weird that arm7 is mapping stuff to 0x27E0000... The GBATEK memory map seems to indicate that the DTCM is ARM9-only. So it seems like arm7 should have no business touching it. Unless it's for some weird IPC stuff?

Reply:
> Interestingly, no$gba shows that 0x27E0000 contains the arm7 code I expect when I switch to nds7, but when I switch back to nds9 it shows me a different set of values

Ah okay, I guess that refutes what I just said. So maybe we don't need to think about the DTCM in that case. If the code matches in arm7 mode, let's keep what you have.

Reply:
> Thanks. Thinking about this more, it's kind of weird that arm7 is mapping stuff to 0x27E0000... The GBATEK memory map seems to indicate that the DTCM is ARM9-only. So it seems like arm7 should have no business touching it. Unless it's for some weird IPC stuff?

It's still weird that it maps to 0x27E0000 in a way, but actually DTCM is separate from the 4MB of main memory right? This means that since arm9 maps 0x27E0000 to DTCM, that specific part of the main memory remains unused, which can be utilized by arm7 for placing its own code.

Reply:
> It's still weird that it maps to 0x27E0000 in a way, but actually DTCM is separate from the 4MB of main memory right? This means that since arm9 maps 0x27E0000 to DTCM, that specific part of the main memory remains unused, which can be utilized by arm7 for placing its own code.

But based on what you said earlier, it seems like arm9 and arm7 are allowed to have different views of what data is at a certain memory address (I guess they have their own memory, or there's some virtual memory thing going on?), so it seems like concerns about what memory is used shouldn't actually matter. So this still feels mysterious to me (though I don't think that should stop us from making the additions as we are here).

Reply:
> though I don't think that should stop us from making the additions as we are here

Agreed, I just have a few more addresses that I need to calculate for unmapped.

> But based on what you said earlier, it seems like arm9 and arm7 are allowed to have different views of what data is at a certain memory address (I guess they have their own memory, or there's some virtual memory thing going on?), so it seems like concerns about what memory is used shouldn't actually matter.

I believe based on what I read [here](https://www.copetti.org/writings/consoles/nintendo-ds/) that the main memory is shared which is where I got this hypothesis, but nowhere does it really talk about the DTCM. Since it's tightly coupled, I almost think it might be hard-coded in the processor perhaps?

Reply:
> Btw do you know how one can do a full memory dump in no$gba? It always says "Huh?" for me whenever I try to dump the whole thing.

It seems like the maximum dump size is 0x1000000, so you probably can't just dump everything at once, unless you manually move to the start of each region and perform separate dumps from there.

About the DTCM stuff: No idea why or how ARM7 uses it, but I know ARM9 holds the stack and the ARM9 interrupt vector there.